### ISSUE TITLE
Values in extension dictionary can potentially be encrypted

### ISSUE THREAD

**Issue by MatthiasValvekens** [pdf-association/pdf-issues#164]
I noticed this one while prototyping the cryptography-related extensions to PDF that are currently in the ISO pipeline (ISO 32003 and 32004). These each come with an extension dictionary to describe the extension, but since some of the values in that dictionary are strings (to wit, **URL** and **ExtensionRevision**), they are (by default) subject to encryption as well!

This is a little awkward when dealing with extensions that _themselves_ add new encryption capabilities to the file format (like ISO 32003, which defines a whole new crypt filter). When dealing with a document using that sort of extension, a processor might simply not understand all of the entries in the extension dictionary, unless it already supported the extension in the first place! And of course, at that point, the processor probably has little use for the contents of the extensions dictionary.

Was that behaviour intended when the **URL** and **ExtensionRevision** entries were added in PDF 2.0?

**Comment by petervwyatt** [comment]
Great insight! The benefit of implementation preceding specification...

I think we could live with an encrypted **ExtensionRevision** since it is a refinement of the concept: "... _provides additional revision information on the extension level being used_."

I see the URL as being more of an issue since that is where an end-user might need to go to understand what they need to do to access a PDF. But this is effectively metadata and PDF already supports unencrypted metadata so maybe some recommendation (informative note?) about duplicating this information for encryption extensions might help???

**Comment by MatthiasValvekens** [comment]
That could indeed be useful in some cases... Do we already have an established mechanism to declare extensions in XMP?

**Comment by petervwyatt** [comment]
AFAIR not for "generic PDF". 
But for PDF/A-1 there is [Technical Note 0009: XMP Extension Schemas in PDF/A-1](https://www.pdfa.org/resource/technical-note-tn-0009-xmp-extension-schemas-in-pdfa-1/)

**Comment by petervwyatt** [comment]
WG8 is making other changes to ISO/TS 32003 (AES-GCM) so there as an opportunity for this to be resolved in that TS. Passing to the PLs to think through a general requirement.

**Comment by petervwyatt** [comment]
PDF 2.0 Unencrypted wrapper documents can be used to encapsulate a PDF using ISO/TS 32003 (AES-GCM) so the UX can be made acceptable, although only the human reading the page of the unencrypted wrapper document will know this.

**Comment by petervwyatt** [comment]
PDF TWG agree that this might be re-considered as an enhancement in a future change to ISO 32000

**Comment by MatthiasValvekens** [comment]
> PDF TWG agree that this might be re-considered as an enhancement in a future change to ISO 32000

To clarify this in light of a previous comment in the thread:

> WG8 is making other changes to ISO/TS 32003 (AES-GCM) so there as an opportunity for this to be resolved in that TS. Passing to the PLs to think through a general requirement.

The reason why we put the ball back in 32k's court is because ISO/TS 32003, _being an extension itself_, cannot meaningfully change the rules on how extensions can be "discovered" by general-purpose PDF processors. If ISO/TS 32003 would contain provisions describing an alternative mechanism to signal its presence, then a PDF processor wouldn't be able to make use of that mechanism unless it was designed with ISO/TS 32003 in mind, and then we're back at square one. Catch-22.

So, long story short: this really needs to be addressed in (a future version of) the core PDF standard, but it may be too big to process as an erratum. For the time being, extensions that augment PDF's encryption capabilities (e.g. ISO/TS 32003) will have to make do with the non-string entries in the extension dictionary.

**Comment by MatthiasValvekens** [comment]
DigSig TWG resolves:

Add NOTE explaining that this is the expected behaviour (wordsmith later).

**Comment by petervwyatt** [comment]
Add NOTE below Table 49:
> NOTE If a developer extension defines a new PDF encryption algorithm then the **URL** and **ExtensionRevision** string values will be encrypted using that new algorithm and will not be meaningful to processors that do not support that developer extension.

**Comment by mkl-public** [comment]
Strictly speaking they only will be encrypted if encryption in the document is set to encrypt the whole document, not e.g. merely attachments. Thus, one may want to re-formulate this to say that the *URL* and *ExtensionRevision* are not automatically exempt from encryption (but may be by setting or nature of the encryption mechanism in question).
